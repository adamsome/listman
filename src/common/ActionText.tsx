/** @jsx jsx */
import { css, jsx } from '@emotion/core'
import React from 'react'
import {
  calcTextHeight,
  calcTextInsetPadding,
  DEFAULT_SIZE_PROP,
  getTextBorderRadius,
  useTheme,
} from '../theming'
import FlexBox from './Flexbox'
import withButtonProps from './with-button-props'

type Props = typeof defaultProps & {
  children: React.ReactNode
  lit?: boolean
  litBG?: string
  hoverBG?: string
  onClick?: (event: React.MouseEvent) => void
}

const defaultProps = {
  ...DEFAULT_SIZE_PROP,
}

const ActionText = (props: Props) => {
  const _props = withButtonProps(props)
  const { lit, litBG, hoverBG, children, size, ...rest } = _props
  const { onClick } = rest
  const click = (event: React.MouseEvent) => {
    if (onClick) {
      event.preventDefault()
      onClick(event)
    }
  }
  const theme = useTheme()
  const fontSize = theme.font.size[size]
  const insetPadding = calcTextInsetPadding(fontSize)
  const height = calcTextHeight(fontSize)
  return (
    <FlexBox
      display="inline-flex"
      justify="center"
      onClick={click}
      css={css`
        font-size: ${fontSize};
        font-weight: ${theme.font.weightBold};
        /** Min width is 2x the inset padding subtracted from the height */
        min-width: calc(${height} - 2 * ${insetPadding});
        border-radius: ${getTextBorderRadius(size)};
        background: ${onClick && lit && theme.bg.bodyDim};
        cursor: ${onClick ? 'pointer' : 'auto'};
        height: ${height};
        line-height: ${height};
        margin-right: ${theme.space.inset};
        padding-left: ${insetPadding};
        padding-right: ${insetPadding};
        outline: none;

        &:hover {
          background: ${onClick && (hoverBG || theme.bg.bodyLit)};
        }
        &:focus {
          box-shadow: 0 0 0 2px ${theme.border.focus}, ${theme.shadow};
        }
      `}
      {...rest}
    >
      {children}
    </FlexBox>
  )
}

ActionText.defaultProps = defaultProps

export default ActionText
