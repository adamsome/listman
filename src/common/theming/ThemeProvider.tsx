/** @jsx jsx */
import { css, jsx } from '@emotion/core'
import { ThemeProvider as EmotionThemeProvider } from 'emotion-theming'
import React, { useEffect, useState } from 'react'
import { getThemeState } from './theme-store'
import THEME_WHITE from './theme-white'
import { Theme } from './theming'

type Props = typeof defaultProps & {
  theme?: Theme
  children: React.ReactNode
}

const defaultProps = {}

const ThemeProvider = (props: Props) => {
  const [currentTheme, setCurrentTheme] = useState(props.theme || THEME_WHITE)

  const handleThemeChange = (next: Theme) => {
    setCurrentTheme(next)
  }

  useEffect(() => {
    const subscription = getThemeState().subscribe(handleThemeChange)
    return () => subscription.unsubscribe()
  }, [])

  const theme = props.theme || currentTheme

  return (
    <EmotionThemeProvider theme={theme}>
      <div
        css={css`
          color: ${theme.body};
          background-color: ${theme.bg.body};
        `}
      >
        {props.children}
      </div>
    </EmotionThemeProvider>
  )
}

ThemeProvider.defaultProps = defaultProps

export default ThemeProvider
