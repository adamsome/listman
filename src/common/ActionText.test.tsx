import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import ActionText from './ActionText'
import ThemeProvider from './theming/ThemeProvider'

const themed = () => (
  <ThemeProvider>
    <ActionText>Test</ActionText>
  </ThemeProvider>
)

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(themed(), div)
  ReactDOM.unmountComponentAtNode(div)
})

it('renders correctly', () => {
  const tree = renderer.create(themed()).toJSON()
  expect(tree).toMatchSnapshot()
})
