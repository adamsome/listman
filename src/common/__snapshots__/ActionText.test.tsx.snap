// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.emotion-1 {
  color: hsl(0,0%,13%);
  background-color: hsl(0,0%,100%);
}

.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: 1.000000rem;
  font-weight: 700;
  min-width: calc(calc(1.000000rem + 0.5rem) - 2 * calc(1.000000rem / 5 + 0.21rem));
  border-radius: 5px;
  cursor: auto;
  height: calc(1.000000rem + 0.5rem);
  line-height: calc(1.000000rem + 0.5rem);
  margin-right: 1rem;
  padding-left: calc(1.000000rem / 5 + 0.21rem);
  padding-right: calc(1.000000rem / 5 + 0.21rem);
  outline: none;
}

.emotion-0:focus {
  box-shadow: 0 0 0 2px hsla(206,71%,83%,0.5),0 1px 2px 0 rgba(0,0,0,0.2),0 2px 4px 0 rgba(0,0,0,0.2);
}

<div
  className="emotion-1"
>
  <div
    className="emotion-0"
    display="inline-flex"
    onClick={[Function]}
  >
    Test
  </div>
</div>
`;
